File: 0001-two-sum/0001-two-sum.py
class Solution(object):
    def twoSumSlow(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        for i in range(0,len(nums)):
            for j in range(i+1,len(nums)):
                if nums[i] + nums[j] == target:
                    return [i,j]

    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        dick={}
        for i in range(0,len(nums)):
            dick[nums[i]]=i
        
        for i in range(0,len(nums)):
            comp = target - nums[i]
            if comp in dick and i != dick[comp]:
                return [i, dick[comp]]
        

File: 0001-two-sum/0001-two-sum.java
import java.util.HashMap;
import java.util.Map;

/**
 * LeetCode 1. Two Sum
 * Java implementation compatible with LeetCode judge.
 */
public class Solution {
    /**
     * Returns indices of the two numbers such that they add up to target.
     * Time: O(n), Space: O(n)
     */
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>(); // value -> index
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement)) {
                return new int[] { map.get(complement), i };
            }
            map.put(nums[i], i);
        }
        throw new IllegalArgumentException("No two sum solution");
    }
}
